File: src/Json/Process/Value.fs
namespace ARCtrl.Json


open Thoth.Json.Core

open ARCtrl
open ARCtrl.Process
open System.IO

module Value = 

    module ISAJson = 

        let encoder (idMap : IDTable.IDTableWrite option) (value : ScalarValue) = 
            match value with
            | ScalarValue.Float f -> 
                Encode.float f
            | ScalarValue.Int i -> 
                Encode.int i
            | ScalarValue.Name s -> 
                Encode.string s
            | ScalarValue.Ontology s -> 
                OntologyAnnotation.ISAJson.encoder idMap s

        let decoder : Decoder<ScalarValue> =
            Decode.oneOf [
                Decode.map ScalarValue.Int Decode.int
                Decode.map ScalarValue.Float Decode.float
                Decode.map ScalarValue.Ontology OntologyAnnotation.ISAJson.decoder
                Decode.map ScalarValue.Name Decode.string           
            ]

File: js/Json/Process/Value.js
import { OntologyAnnotation_ISAJson_decoder, OntologyAnnotation_ISAJson_encoder } from "../OntologyAnnotation.js";
import { string, float, int, map, oneOf } from "../../fable_modules/Thoth.Json.Core.0.4.0/Decode.fs.js";
import { Value } from "../../Core/Value.js";
import { ofArray } from "../../fable_modules/fable-library-js.4.22.0/List.js";

export function encoder(idMap, value) {
    switch (value.tag) {
        case 1:
            return {
                Encode(helpers_1) {
                    return helpers_1.encodeSignedIntegralNumber(value.fields[0]);
                },
            };
        case 3:
            return {
                Encode(helpers_2) {
                    return helpers_2.encodeString(value.fields[0]);
                },
            };
        case 0:
            return OntologyAnnotation_ISAJson_encoder(idMap, value.fields[0]);
        default:
            return {
                Encode(helpers) {
                    return helpers.encodeDecimalNumber(value.fields[0]);
                },
            };
    }
}

export const decoder = oneOf(ofArray([map((Item) => (new Value(1, [Item])), int), map((Item_1) => (new Value(2, [Item_1])), float), map((Item_2) => (new Value(0, [Item_2])), OntologyAnnotation_ISAJson_decoder), map((Item_3) => (new Value(3, [Item_3])), string)]));

